### ISSUE TITLE
feat(p2p-upgrade): Read APK version, bundleId and supported arch before installing

### ISSUE THREAD

**Issue by gmaclennan** [digidem/mapeo-mobile#585]
p2p upgrade currently trusts metadata from a peer to know the version of an APK.

We should check the APK to ensure that it really is the version the metadata says it is, and also check:

- `bundleId` (also called `applicationId`) e.g. `com.mapeo`
- Supported architectures e.g. `arm64-v8a`, `armabi-v7a`

There is an Android Java API to read this data, and that API can be called from React Native e.g. see https://github.com/staltz/dat-installer/blob/master/android/app/src/main/java/com/staltz/datinstaller/PackageInfoModule.java

However, this information is needed by the backend nodejs process, so accessing it via the Android Java API would require a complex back-and-forth via the React Native process.

It would be better to read this information directly from the APK in the nodejs-mobile process. An APK is a Zip archive, and the version and bundleId are stored in the file `AndroidManifest.xml` that is always present in the root of the Zip archive (see https://developer.android.com/guide/topics/manifest/manifest-intro). `AndroidManifest.xml` is encoded as binary XML.

Fortunately there is a tool to read this information: [`app-info-parser`](https://github.com/chenquincy/app-info-parser). This tool also reads the `resources.arsc` file, which includes the app icons, which we don't need to it adds a small memory overhead, but if this is a performance issue in the future we could fork `app-info-parser` and make that parsing separate.

Note that the `package` field of the APK almost always, but not actually always, the same as the `applicationId` see https://developer.android.com/guide/topics/manifest/manifest-element#package I think in our case we can trust that it is the same, but I don't quite understand this difference.

Unfortunately there is no Java API for reading the supported architectures within an APK (at least I can't find one). I think the best way to read the supported architectures of an APK would be to read the contents of the `lib` folder in the APK. It will contain a folder for each supported architecture. I have started work to add a [`readdir()` method to `random-access-zip`](https://github.com/gmaclennan/random-access-zip/pull/9) (also, the name is bad, it should be `zip-fs` since it replicates the `fs` methods for a zipfile).

Ping @noffle 

Depends on #589. Part of #559.

**Comment by gmaclennan** [comment]
Update: I have added `readdir()`, and renamed the module to [`@gmaclennan/zip-fs`](https://www.npmjs.com/package/@gmaclennan/zip-fs) (npm would not allow `zip-fs` because it was too similar to a package named [`zipfs`](https://www.npmjs.com/package/zipfs) which was last published 6 years ago…)

**Comment by hackergrrl** [comment]
> p2p upgrade currently trusts metadata from a peer to know the version of an APK. We should check the APK to ensure that it really is the version the metadata says it is

1. Is this worthwhile (ie. as an attack vector to protect against)? Android won't install an older version, as I understand it, and if we're checking the bundleId we know that it was a build that *we* made, so the worst that could happen is someone advertises a legit APK as `5.0.0` but it's really some newer version instead.

2. EDIT: Actually, what happens if I sign an Android app with the same bundleId as Mapeo set? The signature would be valid and the bundleId would match -- so what stops an attacker from generating such APKs to replace Mapeo?

**Comment by hackergrrl** [comment]
I split the parser work into its own issue: https://github.com/digidem/mapeo-mobile/issues/589. The parent issue of this is https://github.com/digidem/mapeo-mobile/issues/559.

**Comment by gmaclennan** [comment]
> > p2p upgrade currently trusts metadata from a peer to know the version of an APK. We should check the APK to ensure that it really is the version the metadata says it is
> 
> 1. Is this worthwhile (ie. as an attack vector to protect against)? Android won't install an older version, as I understand it, and if we're checking the bundleId we know that it was a build that _we_ made, so the worst that could happen is someone advertises a legit APK as `5.0.0` but it's really some newer version instead.
> 2. EDIT: Actually, what happens if I sign an Android app with the same bundleId as Mapeo set? The signature would be valid and the bundleId would match -- so what stops an attacker from generating such APKs to replace Mapeo?

Good questions!

Scenario: A user has the production Mapeo installed (bundleId `com.mapeo`). A hacker creates a "hacked" version of Mapeo with bundleId `com.mapeo` and signs it with their own key.

Result: Android will refuse to install this over the existing Mapeo installation, because the signatures do not match (it will only install over an existing app with an APK signed with the same private key as the existing app).

Scenario: A user has the production Mapeo installed (bundleId `com.mapeo`). A hacker creates a "hacked" version of Mapeo with bundleId `com.mapeo.hacked` and signs it with their own key.

Result: Android would happily install this if there was no existing app on the device with the bundleId `com.mapeo.hacked`. It could have the same name and icon as the "real" Mapeo app.

---

Since the user is giving permission to Mapeo to "install from untrusted sources", then I think the bundleId check is the most important, because by ensuring we only allow installs of an APK with a bundleId that matches the bundleId of the currently running Mapeo (e.g. for updating itself) then it ensures that the signatures match.

As for the version name and version code, I can't imagine any consequence of sharing an update with incorrect metadata, other than things just working weird. Android does not actually pay any attention to the version name (e.g. `5.2.0`), it only checks the version code when installing, which is an integer. The version code must be >= the version code of the currently installed app.

**Comment by hackergrrl** [comment]
> Since the user is giving permission to Mapeo to "install from untrusted sources", then I think the bundleId check is the most important, because by ensuring we only allow installs of an APK with a bundleId that matches the bundleId of the currently running Mapeo (e.g. for updating itself) then it ensures that the signatures match.

According to what you wrote above, same bundleId does NOT imply same signature, since I could easily make a new Android app with the `com.mapeo` bundleId. This bundleId check would only be useful for the case where someone tries to pass off e.g. the QA version of Mapeo under the Debug discovery key.

> As for the version name and version code, I can't imagine any consequence of sharing an update with incorrect metadata, other than things just working weird. Android does not actually pay any attention to the version name (e.g. 5.2.0), it only checks the version code when installing, which is an integer. The version code must be >= the version code of the currently installed app.

Ok, so we don't need to include code to check the version for this work.

**Comment by gmaclennan** [comment]
Yes what I mean by those scenarios is:

If we check the bundleId is the same as the installed apk, then it can't be installed if signatures do not match

If we don't check it someone could create an app that looks like mapeo (same name and icon) and has a different bundleId, and Android would install it.

**Comment by hackergrrl** [comment]
> If we don't check it someone could create an app that looks like mapeo (same name and icon) and has a different bundleId, and Android would install it.

But wouldn't they just copy Mapeo's bundleId too? (e.g. `com.mapeo`)

**Comment by gmaclennan** [comment]
Android would not allow you to install it (because it has the same bundleId as the current app, but a different signature)

**Comment by gmaclennan** [comment]
Resolved in #595
